core/lib/multivm/src/versions/vm_fast/vm.rs

```diff
@@ -205,7 +205,22 @@ impl<S: ReadStorage, Tr: Tracer> Vm<S, Tr> {
                 }
                 Hook::TxHasEnded => {
                     if let VmExecutionMode::OneTx = execution_mode {
-                        break (last_tx_result.take().unwrap(), false);
+                        // The bootloader may invoke `TxHasEnded` hook without posting a tx result previously. One case when this can happen
+                        // is estimating gas for L1 transactions, if a transaction runs out of gas during execution.
+                        let tx_result = last_tx_result.take().unwrap_or_else(|| {
+                            let tx_has_failed = self.get_tx_result().is_zero();
+                            if tx_has_failed {
+                                let output = VmRevertReason::General {
+                                    msg: "Transaction reverted with empty reason. Possibly out of gas"
+                                        .to_string(),
+                                    data: vec![],
+                                };
+                                ExecutionResult::Revert { output }
+                            } else {
+                                ExecutionResult::Success { output: vec![] }
```

Comment:
Can there be success without getting a result?

Reply:
I'm not sure. I've reproduced functionality from the results tracer for the latest VM:

https://github.com/matter-labs/zksync-era/blob/3fd2fb14e7283c6858731e162522e70051a8e162/core/lib/multivm/src/versions/vm_latest/tracers/result_tracer.rs#L298-L312

Reply:
Looks like the else would be reached if the bootloader runs out of gas after a successful transaction. Apparently that is considered a success.